coreax/solvers/coresubset.py

```diff
@@ -0,0 +1,439 @@
+        :math:`M` points have been selected.
+
+        :param coresubset: The coresubset to refine
+        :param solver_state: Solution state information, primarily used to cache
+            expensive intermediate solution step values.
+        :return: a refined coresubset and relevant intermediate solver state information
+        """
+        x = dataset.data
+        if solver_state is None:
+            gramian_diagonal = jax.vmap(self.kernel.compute_elementwise)(x, x)
+        else:
+            gramian_diagonal = solver_state.gramian_diagonal
+        initial_coresubset = _initial_coresubset(self.coreset_size, dataset)
+        coreset_indices = initial_coresubset.unweighted_indices
+        num_data_points = len(x)
+
+        def _greedy_body(
+            i: int, val: tuple[Array, Array, Array]
+        ) -> tuple[Array, Array, Array]:
+            """RPCholesky iteration - Algorithm 1 of :cite:`chen2023randomly`."""
+            residual_diagonal, approximation_matrix, coreset_indices = val
+            key = jr.fold_in(self.random_key, i)
+            pivot_point = jr.choice(
+                key, num_data_points, (), p=residual_diagonal, replace=False
+            )
+            updated_coreset_indices = coreset_indices.at[i].set(pivot_point)
+            # Remove overlap with previously chosen columns
+            g = (
+                self.kernel.compute(x, x[pivot_point])
+                - (approximation_matrix @ approximation_matrix[pivot_point])[:, None]
+            )
+            updated_approximation_matrix = approximation_matrix.at[:, i].set(
+                jnp.ravel(g / jnp.sqrt(g[pivot_point]))
+            )
+            # Track diagonal of residual matrix and ensure it remains non-negative
+            updated_residual_diagonal = jnp.clip(
+                residual_diagonal - jnp.square(approximation_matrix[:, i]), min=0
+            )
+            if self.unique:
+                # ensures that index selected_pivot_point can't be drawn again in future
+                updated_residual_diagonal = updated_residual_diagonal.at[
+                    pivot_point
+                ].set(0.0)
+            return (
+                updated_residual_diagonal,
+                updated_approximation_matrix,
+                updated_coreset_indices,
+            )
+
+        approximation_matrix = jnp.zeros((num_data_points, self.coreset_size))
+        init_state = (gramian_diagonal, approximation_matrix, coreset_indices)
+        output_state = jax.lax.fori_loop(
+            1, self.coreset_size + 2, _greedy_body, init_state
+        )
+        _, _, updated_coreset_indices = output_state
+        updated_coreset = Coresubset(updated_coreset_indices, dataset)
+        return updated_coreset, RPCholeskyState(gramian_diagonal)
+
+
+def _convert_stein_kernel(
+    x: ArrayLike, kernel: Kernel, score_matching: Union[ScoreMatching, None]
+) -> SteinKernel:
+    """Convert the kernel to a :class:`~coreax.kernel.SteinKernel`."""
```

Comment:
Needs more of an explanation of what the function is doing in its docstring + params?